xbmc/network/osx/darwinICMPPing.m

```diff
@@ -0,0 +1,221 @@
+  struct sockaddr_in callAddress = {0};
+  callAddress.sin_len = sizeof(callAddress);
+  callAddress.sin_family = AF_INET;
+  callAddress.sin_addr.s_addr = inet_addr(host);
+
+  __block NSString *results = kdidStartWithAddress;
+  if (callAddress.sin_addr.s_addr != INADDR_NONE)
+  {
+    [SimplePingClient pingHostAddress:[NSData dataWithBytes:&callAddress length:sizeof(callAddress)]
+             andResultCallback:^(NSString *callbackresults) {
+      results = callbackresults;
+    }];
+  }
+  else
+  {
+    [SimplePingClient pingHostname:[NSString stringWithCString:host encoding:NSUTF8StringEncoding]
+             andResultCallback:^(NSString *callbackresults) {
+      results = callbackresults;
+    }];
+  }
+  
+  float timout_secs = (float)timeout_ms / 1000.0;
+  [[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:timout_secs]];
+
+  if ([results isEqualToString:kdidStartWithAddress])
+  {
+    // host address did not reply, it is not present
+    return -1;
+  }
+  else if ([results isEqualToString:kdidFailWithError])
+  {
+    return -1;
+  }
+  else if ([results isEqualToString:kdidFailToSendPacket])
+  {
+    return -1;
+  }
+  else if ([results isEqualToString:kdidReceiveUnexpectedPacket])
+  {
+    return -1;
+  }
+/*
+  else
+  {
+    NSLog(@"the latency is: %@", results);
+  }
+*/
+  return 0;
+}
+
+int darwinICMPPing(in_addr_t remote_ip, unsigned int timeout_ms)
+{
+  // also checks if host is a name or ip address
+  struct sockaddr_in callAddress = {0};
+  callAddress.sin_len = sizeof(callAddress);
+  callAddress.sin_family = AF_INET;
+  callAddress.sin_addr.s_addr = remote_ip;
+
+  __block NSString *results = kdidStartWithAddress;
+  [SimplePingClient pingHostAddress:[NSData dataWithBytes:&callAddress length:sizeof(callAddress)]
+           andResultCallback:^(NSString *callbackresults) {
+    results = callbackresults;
+  }];
+  
+  float timout_secs = (float)timeout_ms / 1000.0;
+  [[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:timout_secs]];
+
+  if ([results isEqualToString:kdidStartWithAddress])
+  {
+    // host address did not reply, it is not present
+    return -1;
+  }
+  else if ([results isEqualToString:kdidFailWithError])
+  {
+    return -1;
+  }
+  else if ([results isEqualToString:kdidFailToSendPacket])
+  {
+    return -1;
+  }
+  else if ([results isEqualToString:kdidReceiveUnexpectedPacket])
+  {
+    return -1;
+  }
+/*
+  else
+  {
+    NSLog(@"the latency is: %@", results);
+  }
+*/
+  return 0;
+}
+
+
+//********************************************************************************************
+@interface SimplePingClient()
+{
+  SimplePing* _pingClient;
+  NSDate* _dateReference;
+}
+@property(nonatomic, strong) void(^resultCallback)(NSString* result);
+@end
+
+@implementation SimplePingClient
+
++(void)pingHostname:(NSString*)hostName andResultCallback:(void(^)(NSString* result))result
+{
+  static SimplePingClient* singletonPC = nil;
```

Comment:
Is this singleton catching any replies for the duration of the ping? If ping is done by multiple threads is it asserted that the incoming replies is returned only to the correct caller?
(Disclaimer , I dont read applish)